Client/game_sa/CWeaponSA.cpp

```diff
@@ -264,145 +116,89 @@ bool CWeaponSA::FireBullet(CEntity* pFiringEntity, const CVector& vecOrigin, con
+            if (pGame->m_pPostWeaponFireHandler && firingPlayerPed)
                 pGame->m_pPostWeaponFireHandler();
 
             pGame->GetWorld()->IgnoreEntity(nullptr);
-
             return true;
         }
 
         default:
             break;
     }
+
     return false;
 }
 
-bool CWeaponSA::FireInstantHit(CEntity* pFiringEntity, const CVector* pvecOrigin, const CVector* pvecMuzzle, CEntity* pTargetEntity, const CVector* pvecTarget,
-                               const CVector* pvec, bool bFlag1, bool bFlag2)
+bool CWeaponSA::GenerateDamageEvent(CPed* ped, CEntity* responsible, eWeaponType weaponType, int damagePerHit, ePedPieceTypes hitZone, std::uint8_t dir) const
 {
-    bool  bReturn;
-    DWORD dwEntityInterface = 0;
-    if (pFiringEntity)
-        dwEntityInterface = (DWORD)pFiringEntity->GetInterface();
-    DWORD dwTargetInterface = 0;
-    if (pTargetEntity)
-        dwTargetInterface = (DWORD)pTargetEntity->GetInterface();
-    DWORD dwThis = (DWORD)m_pInterface;
-    DWORD dwFunc = FUNC_CWeapon_FireInstantHit;
-    _asm
-    {
-        mov     ecx, dwThis
-        push    bFlag2
-        push    bFlag1
-        push    pvec
-        push    pvecTarget
-        push    dwTargetInterface
-        push    pvecMuzzle
-        push    pvecOrigin
-        push    dwEntityInterface
-        call    dwFunc
-        mov     bReturn, al
-    }
-    return bReturn;
+    if (!ped || !m_interface)
+        return false;
+
+    return m_interface->GenerateDamageEvent(ped, responsible, weaponType, damagePerHit, hitZone, dir);
 }
 
-bool CWeaponSA::ProcessLineOfSight(const CVector* vecStart, const CVector* vecEnd, CColPoint** colCollision, CEntity** CollisionEntity,
-                                   const SLineOfSightFlags flags, SLineOfSightBuildingResult* pBuildingResult, eWeaponType weaponType,
-                                   CEntitySAInterface** pEntity)
+bool CWeaponSA::ProcessLineOfSight(const CVector* vecStart, const CVector* vecEnd, CColPoint** colCollision, CEntity** collisionEntity, const SLineOfSightFlags& flags, SLineOfSightBuildingResult* buildingResult, const eWeaponType& weaponType, CEntitySAInterface** entity)
 {
-    DWORD dwFunction = FUNC_CBirds_CheckForHit;
-    _asm
-    {
-        push vecEnd
-        push vecStart
-        call dwFunction
-    }
-    dwFunction = FUNC_CShadows_CheckForHit;
-    _asm
-    {
-        push vecEnd
-        push vecStart
-        call dwFunction
-    }
-    bool bReturn = pGame->GetWorld()->ProcessLineOfSight(vecStart, vecEnd, colCollision, CollisionEntity, flags, pBuildingResult);
-    _asm
-    {
-        add esp, 10h
-    }
+    // Call CBirds::HandleGunShot
+    ((void(__cdecl*)(CVector*, CVector*))FUNC_CBirds_CheckForHit)(const_cast<CVector*>(vecStart), const_cast<CVector*>(vecEnd));
+
+    // Call CShadows::GunShotSetsOilOnFire
+    ((void(__cdecl*)(CVector*, CVector*))FUNC_CShadows_CheckForHit)(const_cast<CVector*>(vecStart), const_cast<CVector*>(vecEnd));
```

Comment:
Could you define these function parameters as const instead of using const_cast?

Reply:
If agreed on, please address in a follow up PR, because there are schedule reasons to merge now.